apitools/base/py/transfer.py

```diff
@@ -740,9 +740,15 @@ def ConfigureRequest(self, upload_config, http_request, url_builder):
             # cannot happen.
             if self.__gzip_encoded:
                 http_request.headers['Content-Encoding'] = 'gzip'
-                body_buffer = six.BytesIO(http_request.body)
-                body, _, _ = compression.CompressStream(body_buffer)
-                http_request.body = body
+                # Turn the body into a stream so that we can compress it, then
+                # read the compressed bytes.  In the event of a retry (e.g. if
+                # our access token has expired), we need to be able to re-read
+                # the body, which we can't do with a stream. So, we consume the
+                # bytes from the stream now and store them in a re-readable
+                # bytes container.
+                http_request.body = (
+                    compression.CompressStream(
+                        six.BytesIO(http_request.body))[0].read())
```

Comment:
This reads the entire contents of the stream into memory, whereas the CompressStream/StreamingBuffer classes were designed specifically to avoid that.  Should we instead fix this at the forced retry level?  (i.e., that should not be calling read() without rewinding the stream or buffering the bytes that it reads)?

Reply:
The comment on L 736 mentions that "Both multipart and media request uploads will read the entire stream into memory" -- so this doesn't really change the behavior for the simple upload case, it just does the whole read now rather than later... right?  Or am I reading this incorrectly?

Reply:
If we were calling CompressStream with some arguments that limited the length (like we do elsewhere, but not for simple media and multipart requests here), I'd agree with your point... but calling it with no argument for the `length` parameter will just read in the whole stream -- it writes the entire contents of the input stream to the output "stream" (the underlying `StreamingBuffer`), so the whole thing is just sitting there in memory waiting for someone to read from the underlying `StreamingBuffer`.  Those reads may be performed incrementally, but they'll just knock bytes off of the underlying buffer with each read.  The only benefit we lose by converting this directly to a string is not being able to do that incremental reading that slowly reduces the memory used as the bytes are read off the underlying `StreamingBuffer`... and I'd argue that for large files where this actually matters, users should be utilizing the RESUMABLE strategy, rather than SIMPLE.

Thoughts?

Reply:
Thanks - I agree; for SIMPLE this is fine.